File: android/app/src/main/java/com/elmenture/luuk/base/BaseActivity.java
package com.elmenture.luuk.base;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;

import com.elmenture.luuk.R;
import com.elmenture.luuk.User;
import com.elmenture.luuk.utils.LogUtils;
import com.facebook.AccessToken;
import com.facebook.CallbackManager;
import com.facebook.login.LoginManager;
import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;

public abstract class BaseActivity extends AppCompatActivity implements BaseActivityView {

    protected LogUtils logUtils = new LogUtils(this.getClass());
    protected GoogleSignInClient mGoogleSignInClient;
    protected CallbackManager callbackManager;

    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        configureGoogleSignIn();
        configureFacebookSignin();
    }

    private void configureFacebookSignin() {
        callbackManager = CallbackManager.Factory.create();
        AccessToken accessToken = AccessToken.getCurrentAccessToken();

        boolean isLoggedIn = accessToken != null && !accessToken.isExpired();
        if (isLoggedIn) {
            if(!User.hasSignedInUser()){
                LoginManager.getInstance().logOut();
            }
        }
    }

    @Override
    public void addFragment(@NonNull Fragment fragment, @NonNull String tag){
        FragmentTransaction fragmentTransaction = getSupportFragmentManager().beginTransaction();
        fragmentTransaction.add(R.id.layout_fragment_container, fragment, tag);
        fragmentTransaction.addToBackStack(tag).commitAllowingStateLoss();
    }

    @Override
    public void replaceFragment(@NonNull Fragment fragment, @NonNull String tag){
        FragmentTransaction fragmentTransaction = getSupportFragmentManager().beginTransaction();
        fragmentTransaction.replace(R.id.layout_fragment_container, fragment, tag).commitAllowingStateLoss();
    }

    private void configureGoogleSignIn() {
        // Configure sign-in to request the user's ID, email address, and basic
        // profile. ID and basic profile are included in DEFAULT_SIGN_IN.
        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.server_client_id))
                .requestEmail()
                .build();

        // Build a GoogleSignInClient with the options specified by gso.
        mGoogleSignInClient = GoogleSignIn.getClient(this, gso);

        // Check for existing Google Sign In account, if the user is already signed in
        // the GoogleSignInAccount will be non-null.
        GoogleSignInAccount account = GoogleSignIn.getLastSignedInAccount(this);
        if (account != null) {
            if(!User.hasSignedInUser()){
                mGoogleSignInClient.signOut();
            }
        }
    }
}

File: android/app/src/main/java/com/elmenture/luuk/base/BaseActivity.kt
package com.elmenture.luuk.base

import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.Fragment
import com.elmenture.luuk.R
import com.elmenture.luuk.base.repositories.RemoteRepository
import com.google.android.material.snackbar.Snackbar
import utils.LogUtils
import views.ProgressDialog


abstract class BaseActivity : AppCompatActivity(), BaseActivityView {
    @JvmField
    protected var logUtils = LogUtils(this.javaClass)

    private var loader: DialogFragment? = null
    private var snackbar: Snackbar? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        observeGlobalEvents()
    }

    override fun addFragment(fragment: Fragment, tag: String) {
        val fragmentTransaction = supportFragmentManager.beginTransaction()
        fragmentTransaction.add(R.id.layout_fragment_container, fragment, tag)
        fragmentTransaction.addToBackStack(tag).commitAllowingStateLoss()
    }

    override fun replaceFragment(fragment: Fragment, tag: String) {
        val fragmentTransaction = supportFragmentManager.beginTransaction()
        fragmentTransaction.replace(R.id.layout_fragment_container, fragment, tag)
            .commitAllowingStateLoss()
    }

    private fun observeGlobalEvents() {
        RemoteRepository.blockUserInteraction.observe(this) { blockUser ->
            handleProgressDialog(blockUser)
        }
    }

    private fun handleProgressDialog(blockUser: Boolean) {
        if (blockUser) {
            showProgressDialog()
        } else {
            hideProgressDialog()
        }

    }

    fun showProgressDialog() {
        if (loader == null)
            loader = ProgressDialog.newInstance()
        if (loader!!.isAdded) {
            loader?.dismissAllowingStateLoss()
        }
        loader?.show(supportFragmentManager, ProgressDialog.TAG)
    }

    fun hideProgressDialog() {
        loader?.dismissAllowingStateLoss()
    }

    fun showMessage(view: View, isSuccessFullAction: Boolean, message: String? = null) {
        snackbar?.dismiss()
        var snackMsg = message
        if (snackMsg.isNullOrEmpty()) {
            snackMsg = "Action Failed. Try Again"
            if (isSuccessFullAction) {
                snackMsg = "Action Completed Successfully"
            }
        }

        snackbar = Snackbar.make(view, snackMsg, Snackbar.LENGTH_INDEFINITE)
        val textView =
            snackbar?.view?.findViewById<TextView>(com.google.android.material.R.id.snackbar_text)
        textView?.gravity = Gravity.CENTER
        textView?.setLines(2);
        textView?.setTextColor(resources.getColor(R.color.luuk_yellow))
        snackbar?.setAction("OK") { snackbar?.dismiss() }
        snackbar?.show()
    }


    override fun onBackPressed() {
        super.onBackPressed()
        val fragment = supportFragmentManager.findFragmentById(R.id.layout_fragment_container)
        fragment?.let {
            if (it is BaseFragment)
                it.onResumeFromBackstack()
        }
    }
}